sdk/core/Azure.Core/src/Shared/TaskExtensions.cs

```diff
@@ -133,5 +151,69 @@ private static void VerifyTaskCompleted(bool isCompleted)
             public void Dispose() => _asyncEnumerator.DisposeAsync().EnsureCompleted();
 #pragma warning restore AZC0107 // Do not call public asynchronous method in synchronous scope.
         }
+
+        public readonly struct WithCancellationAwaitable<T>
+        {
+            private readonly CancellationToken _cancellationToken;
+            private readonly ConfiguredTaskAwaitable<T> _awaitable;
+
+            public WithCancellationAwaitable(Task<T> task, CancellationToken cancellationToken)
+            {
+                _awaitable = task.ConfigureAwait(false);
+                _cancellationToken = cancellationToken;
+            }
+
+            public WithCancellationAwaiter<T> GetAwaiter() => new WithCancellationAwaiter<T>(_awaitable.GetAwaiter(), _cancellationToken);
+        }
+
+        public readonly struct WithCancellationAwaiter<T> : ICriticalNotifyCompletion
```

Comment:
Why don't we implement this via an async method that takes a task?

Reply:
Not sure I understand your question. Async method that takes a task is `WithCancellation` above.

Reply:
I mean why do we need a custom awaiter to get this feature.

Wouldn't the following be enough:
```
public static ValueTask<T> WithCancellation<T>(this Task<T> task, CancellationToken cancellationToken)
{
cancellationToken.ThrowIfCancellationRequested();
var result = await task;
cancellationToken.ThrowIfCancellationRequested();
return result;
}
``

Reply:
If `cancellationToken` is canceled after `await` is started but before `task` is completed, we shouldn't wait for that task, we should continue with throwing `OperationCanceledException`.